R/sql_translate_env.R

```diff
@@ -68,7 +68,18 @@ sql_translate_env.PrestoConnection <- function(con) {
       pmin = sql_prefix("LEAST"),
       is.finite = sql_prefix("IS_FINITE"),
       is.infinite = sql_prefix("IS_FINITE"),
-      is.nan = sql_prefix("IS_NAN")
+      is.nan = sql_prefix("IS_NAN"),
+      `[[` = function(x, i) {
+        i <- dplyr::enexpr(i)
```

Comment:
We still support dplyr 0.7.0 which unfortunately does not seem to have `enexpr` (https://github.com/tidyverse/dplyr/blob/v0.7.0/NAMESPACE). The earliest is 0.7.5. Any chance you could find an alternative?

Reply:
Sure, I've replaced it by `substitute()` and added additional tests for remote evaluation of both `x` and `i`.

Reply:
Great, thanks.

Actually, could you also explain why we need the substitute?

Reply:
The issue is whether we want to throw an error if the index `i` is unquoted. Let's examine the behavior without the `substitute()`, and also examine `i` by making the following change:

```r
`[[` = function(x, i) {
  cat(str(i))
  # i <- substitute(i)
  if (is.character(i)) {
  ...
```

We'll rebuild the package and run the below reprex. When we pass an unquoted index, i.e., `a` versus `"a"`, it will be passed down as a SQL object. Because this also has class `character`, `is.character()` will be `TRUE` and we will attempt to translate it, resulting in `"x"[["a"]]`.

```r
library(RPresto)
source("tests/testthat/utilities.R")
translate_sql <- RPresto:::dbplyr_compatible('translate_sql')
s <- setup_mock_dplyr_connection()[['db']]
translate_sql(x[[a]], con=s[['con']])
#> Classes 'sql', 'character'  chr "\"a\""
#> <SQL> "x"["a"]
```

This seems like a bug waiting to happen. To avoid it, we can use `substitute()` to turn the index into (in this case) a `symbol`, which fails the `is.character()` test and throws an error. Undoing the modification and re-running the reprex, we get

```r
library(RPresto)
source("tests/testthat/utilities.R")
translate_sql <- RPresto:::dbplyr_compatible('translate_sql')
s <- setup_mock_dplyr_connection()[['db']]
translate_sql(x[[a]], con=s[['con']])
#> Error in x[[a]]: Error: `i` must be character or numeric
```

I think the version with `substitute()` is cleaner, what do you think? If you agree, I can add a test to make this clear.

Reply:
Oh I see. Presto allows dynamic indexing for arrays and maps though:

    SELECT
        x["a"],
        y[a]
    FROM (
        VALUES
            (ARRAY[1, 2, 3], MAP(ARRAY[2, 4], ARRAY['a', 'b']), 2),
            (ARRAY[4, 5], NULL, 1),
            (ARRAY[6, 7], MAP(ARRAY[5, 6], ARRAY['x', 'y']), 2)
    ) AS t(x, y, a)

So `x[a]` is not necessarily invalid.

Reply:
Good point. What if we do something like this?

```r
`[[` = function(x, i) {
  i <- substitute(i)
  i_quoted <- switch(
    typeof(i),
    character = dbplyr::escape(i, con = con),
    double = as.integer(i),
    integer = i,
    symbol = dbplyr::ident_q(deparse(i))
  )
  if (!is.null(i_quoted)) {
    dbplyr::build_sql(x, "[", i_quoted, "]")
  } else {
    stop("Error: `i` must be character, numeric, or a symbol")
  }
}
```

Then we can allow both of the following (plus the numeric cases):

```r
library(RPresto)
source('tests/testthat/utilities.R')
translate_sql <- RPresto:::dbplyr_compatible('translate_sql')
s <- setup_mock_dplyr_connection()[['db']]
translate_sql(x[[a]], con=s[['con']])
#> <SQL> "x"[a]
translate_sql(x[['a']], con=s[['con']])
#> <SQL> "x"['a']
```

Reply:
I think that is too complicated. Here's what I have done:

    `[[` = function(x, i) {
       # Convenience for array indices. We check for equality in conversion since double keys are allowed for maps.
       if (is.numeric(i) && isTRUE(all.equal(i, as.integer(i)))) {
         i <- as.integer(i)
       }
       build_sql(x, "[", i, "]")
     }

And the test case:

    library('RPresto')
    library('dplyr')

    source('tests/testthat/utilities.R')
    t <- (
      tbl(
        setup_live_dplyr_connection()[['db']],
        sql("SELECT * FROM (VALUES
          (
            ARRAY['1_first', '1_second', '1_third', '1_fourth'],
            MAP(ARRAY[1, 2, 3, 4], ARRAY['map_1', 'map_2', 'map_3', 'map_4']),
            2.0
          ),
          (
            ARRAY['2_first', '2_second', '2_third', '2_fourth'],
            MAP(ARRAY[1.5, 2.5, 3.5], ARRAY['map_1.5', 'map_2.5', 'map_3.5']),
            1.0
          )
        ) AS data(arr, m, idx)")
      )
    )
    
    query <- (
      t
      %>% mutate(
        array_function_static=arr[[as.integer(POW(2, 2))]],
        array_function_dynamic=arr[[as.integer(POW(idx, 2))]],
        map_function_static=m[[as.integer(POW(2, 2))]],
        map_function_dynamic=m[[as.integer(POW(idx, 2))]],
        array_direct_static_numeric=arr[[1]],
        array_direct_static_integer=arr[[1L]],
        map_direct_static_numeric=m[[1]],
        map_direct_static_numeric_half=m[[1.5]],
        map_direct_static_integer=m[[1L]],
        array_direct_dynamic_integer=arr[[as.integer(idx)]],
        map_direct_dynamic_numeric=m[[idx]],
        map_direct_dynamic_numeric_half=m[[idx + 0.5]],
        map_direct_dynamic_integer=m[[as.integer(idx)]]
      )
    )
    
    show_query(query)
    collect(query)

Reply:
I agree this is a lot simpler. Do you want to throw an error for any type of index, e.g., `x[[TRUE]]`, or just try to translate everything and let the error get thrown elsewhere in the call stack (possibly by Presto)?

Reply:
I don't think we can validate indices in R land easily. `map`s are quite flexible in terms of key types they allow. `boolean`s as in your example would be valid, for instance. It will have to be in Presto land.

Reply:
PR now reflects the simplified translation, plus the test cases you give above.